File: leetcode/1_two_sum/solution.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # one-pass hash table
        hash_table = dict()
        for idx, num in enumerate(nums):
            complement = target - num
            if complement in hash_table and hash_table[complement] != idx:
                return [hash_table[complement], idx]
            hash_table[num] = idx
        

if __name__ == "__main__":
    sol = Solution()
    assert sol.twoSum([2,7,11,15], 9) == [0, 1]
    assert sol.twoSum([3,2,4], 6) == [1, 2]
    assert sol.twoSum([3,3], 6) == [0, 1]

File: leetcode/1_two_sum/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int idx = 0; idx < nums.length; idx++) {
            int complement = target - nums[idx];
            if (map.containsKey(complement)) {
                return new int[]{map.get(complement), idx};
            }
            map.put(nums[idx], idx);
        }
        throw new IllegalArgumentException("No Two Sum Solution");
    }
}